### ISSUE TITLE
dataset path when training custom data

### ISSUE THREAD

**Issue by BeomyoungKim** [ultralytics/yolov5#7495]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

<img width="486" alt="스크린샷 2022-04-20 오후 6 30 24" src="https://user-images.githubusercontent.com/60453970/164197799-5d9fee03-9fad-412a-b17b-ac62e8b08ab2.png">
<img width="304" alt="스크린샷 2022-04-20 오후 6 31 29" src="https://user-images.githubusercontent.com/60453970/164198535-9e9204b8-4dc3-4a58-a37e-64db162f282d.png">

This is my directory structure
and I train my custom data using pypi yolov5.
In yolov5~=6.0.6 it find my custom dataset well 
but when I update to 6.1 version it check only yolov5 lib directory

where should I put my custom dataset when using pypi yolov5 6.1 version

**Comment by glenn-jocher** [comment]
@BeomyoungKim 👋 Hello! Thanks for asking about YOLOv5 🚀 **dataset formatting**. To train correctly your data must be in YOLOv5 format. Please see our [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading) tutorial for full documentation on dataset setup and all steps required to start training your first model. A few excerpts from the tutorial:

### 1.1 Create dataset.yaml

[COCO128](https://www.kaggle.com/datasets/ultralytics/coco128) is an example small tutorial dataset composed of the first 128 images in [COCO](http://cocodataset.org/#home) train2017. These same 128 images are used for both training and validation to verify our training pipeline is capable of overfitting. [data/coco128.yaml](https://github.com/ultralytics/yolov5/blob/master/data/coco128.yaml), shown below, is the dataset config file that defines 1) the dataset root directory `path` and relative paths to `train` / `val` / `test` image directories (or *.txt files with image paths), 2) the number of classes `nc` and 3) a list of class `names`:
```yaml
# Train/val/test sets as 1) dir: path/to/imgs, 2) file: path/to/imgs.txt, or 3) list: [path/to/imgs1, path/to/imgs2, ..]
path: ../datasets/coco128  # dataset root dir
train: images/train2017  # train images (relative to 'path') 128 images
val: images/train2017  # val images (relative to 'path') 128 images
test:  # test images (optional)

# Classes
nc: 80  # number of classes
names: [ 'person', 'bicycle', 'car', 'motorcycle', 'airplane', 'bus', 'train', 'truck', 'boat', 'traffic light',
         'fire hydrant', 'stop sign', 'parking meter', 'bench', 'bird', 'cat', 'dog', 'horse', 'sheep', 'cow',
         'elephant', 'bear', 'zebra', 'giraffe', 'backpack', 'umbrella', 'handbag', 'tie', 'suitcase', 'frisbee',
         'skis', 'snowboard', 'sports ball', 'kite', 'baseball bat', 'baseball glove', 'skateboard', 'surfboard',
         'tennis racket', 'bottle', 'wine glass', 'cup', 'fork', 'knife', 'spoon', 'bowl', 'banana', 'apple',
         'sandwich', 'orange', 'broccoli', 'carrot', 'hot dog', 'pizza', 'donut', 'cake', 'chair', 'couch',
         'potted plant', 'bed', 'dining table', 'toilet', 'tv', 'laptop', 'mouse', 'remote', 'keyboard', 'cell phone',
         'microwave', 'oven', 'toaster', 'sink', 'refrigerator', 'book', 'clock', 'vase', 'scissors', 'teddy bear',
         'hair drier', 'toothbrush' ]  # class names
```


### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```

<p align="center"><img width="700" src="https://user-images.githubusercontent.com/26833433/134436012-65111ad1-9541-4853-81a6-f19a3468b75f.png"></p>
</details>


Good luck 🍀 and let us know if you have any other questions!

**Comment by BeomyoungKim** [comment]
<img width="775" alt="스크린샷 2022-04-21 오전 10 50 23" src="https://user-images.githubusercontent.com/60453970/164355647-5127f637-b950-4192-b1ed-df643f13ee1e.png">
I find solution. By setting absolute path, it can find my own custom dataset when using pypi yolov5 6.1.x

**Comment by glenn-jocher** [comment]
@BeomyoungKim that's great to hear! 🔍 Setting absolute paths can indeed help YOLOv5 locate custom datasets when using the 6.1.x version. Thank you for sharing your solution 🙌. Let us know if you have any other questions or if there's anything else we can assist you with.

**Comment by laviRZ** [comment]
> Setting absolute paths can indeed help

Is there a way to go about this without absolute paths? Is there an environment variable that can point to the datasets path?

**Comment by glenn-jocher** [comment]
@laviRZ Currently, YOLOv5 does not utilize environment variables for custom dataset paths. However, setting up your dataset with relative paths, as described in the [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading/) tutorial, should work seamlessly. Let us know if you encounter any issues or need further assistance. We're here to help!

**Comment by laviRZ** [comment]
@glenn-jocher the paths in my yaml (generated by roboflow) look like this:
```
test: ../test/images
train: 2023-Vision-5/train/images
val: 2023-Vision-5/valid/images
```
and the project structure is 
basedir/yolov5
basedir/datasets/2023-Vision-5/train/images

when running train.py, it looks for the files in basedir/yolov5/2023-Vision-5/train/images

**Comment by glenn-jocher** [comment]
@laviRZ Thanks for sharing the details. It looks like the relative paths in your YAML file are not being resolved correctly. In this case, you can try using the absolute paths directly in your YAML file. For example, you can modify the paths in your YAML file to use absolute paths like this:
```
test: /absolute/path/to/your/dataset/test/images
train: /absolute/path/to/your/dataset/2023-Vision-5/train/images
val: /absolute/path/to/your/dataset/2023-Vision-5/valid/images
```
This should help ensure that the paths are resolved correctly regardless of the current working directory. Let me know if you have any more questions or if there's anything else I can assist you with.
